CellDataCoupler/Classes/Base.swift

```diff
@@ -69,35 +69,71 @@ open class CellCouplerSection {
 }
 
 open class CouplerFactory {
-    public typealias CouplerFetch = ((Int) -> BaseCellCoupler?)
+    public typealias CouplerFetch = ((Int) -> BaseCellCoupler)
+    public typealias MultiCouplerFetch = ((Int) -> [BaseCellCoupler])
     
     private var count: Int
-    private var couplerFetch: CouplerFetch
-    private var cache: [Int: BaseCellCoupler]?
-    
-    public init(count: Int, couplerFetch: @escaping CouplerFetch, cached: Bool = true) {
-        self.count = count
-        self.couplerFetch = couplerFetch
-        cache = cached ? [:] : nil
+    private var countPerIteration: Int
+    private var couplerFetch: MultiCouplerFetch
+    private var cached: Bool
+    private var cache = [Int: [BaseCellCoupler]]()
+        
+    public init(iterationCount: Int, couplerCountPerIteration: Int, multiCouplerFetch: @escaping MultiCouplerFetch, cached: Bool = true) {
+        self.count = iterationCount
+        self.countPerIteration = couplerCountPerIteration
+        self.couplerFetch = multiCouplerFetch
+        self.cached = cached
     }
     
+    public convenience init(count: Int, couplerFetch: @escaping CouplerFetch, cached: Bool = true) {
+        let fetch: MultiCouplerFetch = { (index) -> [BaseCellCoupler] in
+            return [couplerFetch(index)]
+        }
+        
+        self.init(iterationCount: count, couplerCountPerIteration: 1, multiCouplerFetch: fetch, cached: cached)
+    }
+
     internal convenience init(couplers: [BaseCellCoupler]) {
         self.init(count: couplers.count, couplerFetch: { (index) -> BaseCellCoupler in
             return couplers[index]
         }, cached: false)
     }
     
     public func numberOfItems() -> Int {
-        return count
+        return count * countPerIteration
+    }
+    
+    public func getItem(for index: Int) -> BaseCellCoupler {
+        return getItem(for: index, forceCache: false)
     }
     
-    public func getItem(for index: Int) -> BaseCellCoupler? {
-        if let cache = cache, let cachedItem = cache[index] {
-            return cachedItem
+    internal func getItem(for index: Int, forceCache: Bool) -> BaseCellCoupler {
```

Comment:
But `cellForRowAtIndexPath` is non-optional.  We have to provide a cell.  Here's what I was doing before:

        guard let coupler = sections[indexPath.section].factory.getItem(for: indexPath.row) else {
            let errorCell = UITableViewCell()	
            errorCell.textLabel?.text = "CellDataCoupler Error - No Coupler Provided!"	
            return errorCell	
        }

This isn't something we can avoid.

Before, I was allowing the coupler fetch to return a nil coupler, which doesn't make sense, since I was displaying an ugly and invalid cell if the user did give me a null coupler.

I instead refactored this to require the coupler fetch to have a non-nil coupler:

public typealias CouplerFetch = ((Int) -> BaseCellCoupler)

However, with this new arrayed coupler fetch:

    public typealias MultiCouplerFetch = ((Int) -> [BaseCellCoupler])

There's still an opportunity for the user to not give me enough couplers in their array, so it's still something I had to deal with.  Instead of punishing the developer by slapping in an ugly cell into their view (which could get released into prod!) I instead print a descriptive message into the console, and give the tableview an "empty" cell with zero height.  This  makes it essentially invisible.

I also moved this empty cell state into the CouplerFactory, since its this classes' fault I have to deal with this issue anyways, and shouldn't bleed into other classes that don't allow nil coupler/cells.

Reply:
That's a fair point.